File: Questions/PizzaBuilder.py
# Create enums for customizing the pizza
from abc import ABC, abstractmethod
from enum import Enum


class Size(Enum):
    SMALL = "small"
    MEDIUM = "medium"
    LARGE = "large"


class Crust(Enum):
    THIN = "thin"
    THICK = "thick"


class Sauce(Enum):
    TOMATO = "tomato"
    BBQ = "bbq"


class Topping(Enum):
    JALAPENO = "jalapeno"
    OLIVES = "olives"
    PINEAPPLE = "pineapple"
    MUSHROOMS = "mushrooms"


# Define a concrete class for pizza
class Pizza:
    def __init__(self):
        # Individual attributes
        self._size: Size = None
        self._crust: Crust = None
        self._sauce: Sauce = None
        self._toppings: list[Topping] = []

    def __str__(self):
        return (
            f"Pizza(size={self.size}, crust={self.crust}, "
            f"sauce={self.sauce}, toppings={self.toppings})"
        )

    # Create getter and setters for size
    @property
    def size(self):
        return self._size

    @size.setter
    def size(self, size: Size):
        self._size = size

    # Create getter and setter for crust
    @property
    def crust(self):
        return self._crust

    @crust.setter
    def crust(self, crust: Crust):
        self._crust = crust

    # Create getter and setter for sauce
    @property
    def sauce(self):
        return self._sauce

    @sauce.setter
    def sauce(self, sauce: Sauce):
        self._sauce = sauce

    # Create getter and setter for toppings
    @property
    def toppings(self):
        return self._toppings

    @toppings.setter
    def toppings(self, toppings: list[Topping]):
        self._toppings = toppings


# Create the Builder pattern for quickly building custom pizzas
class Builder(ABC):
    @abstractmethod
    def set_size(self, size: Size):
        pass

    @abstractmethod
    def set_crust(self, crust: Crust):
        pass

    @abstractmethod
    def set_sauce(self, sauce: Sauce):
        pass

    @abstractmethod
    def add_topping(self, topping: Topping):
        pass

    @abstractmethod
    def build(self):
        pass


# Define a concrete Pizza builder class
class PizzaBuilder(Builder):
    def __init__(self):
        self.pizza = Pizza()

    def set_size(self, size: Size):
        self.pizza.size = size
        return self

    def set_crust(self, crust: Crust):
        self.pizza.crust = crust
        return self

    def set_sauce(self, sauce: Sauce):
        self.pizza.sauce = sauce
        return self

    def add_topping(self, topping: Topping | list[Topping]):
        # If topping is a list, add the entire list
        if isinstance(topping, list):
            self.pizza.toppings.extend(topping)
        else:
            self.pizza.toppings.append(topping)
        return self

    def build(self):
        return self.pizza


# Test the system
pizza_builder = PizzaBuilder()

my_pizza = (PizzaBuilder()
            .set_size(Size.LARGE)
            .set_crust(Crust.THIN)
            .set_sauce(Sauce.TOMATO)
            .add_topping([Topping.JALAPENO, Topping.OLIVES])
            .build())

print(my_pizza)

File: Java/Questions/PizzaBuilder.java

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// Creating enums for pizza attributes
enum Size {
    SMALL, MEDIUM, LARGE
}

enum Crust {
    THIN, THICK
}

enum Sauce {
    TOMATO, BBQ
}

enum Topping {
    JALAPENO, OLIVES, PINEAPPLE, MUSHROOMS
}

// Creating a concrete pizza class
class Pizza {
    // Variables to store the pizza attributes
    private Size size;
    private Crust crust;
    private Sauce sauce;
    private List<Topping> toppings;

    // Constructor to initialize the pizza 
    public Pizza() {
        this.toppings = new ArrayList<>();
    }

    // Getter and setter for the pizza's size
    public Size getSize() {
        return size;
    }

    public void setSize(Size size) {
        this.size = size;
    }

    // Getter and setter for the pizza's crust
    public Crust getCrust() {
        return crust;
    }

    public void setCrust(Crust crust) {
        this.crust = crust;
    }

    // Getter and setter for the pizza's sauce
    public Sauce getSauce() {
        return sauce;
    }

    public void setSauce(Sauce sauce) {
        this.sauce = sauce;
    }

    // Getter and setter for the pizza's toppings
    public List<Topping> getToppings() {
        return toppings;
    }

    public void setToppings(List<Topping> toppings) {
        this.toppings = toppings;
    }

    // Method to pretty print the pizza
    @Override
    public String toString() {
        return "Pizza(size=" + size +
                ", crust=" + crust +
                ", sauce=" + sauce +
                ", toppings=" + toppings + ")";
    }
}

// Create the PizzaBuilder concrete class
public class PizzaBuilder {
    // Creating private pizza variable
    private final Pizza pizza;

    // Constructor to initialize PizzaBuilder
    public PizzaBuilder() {
        this.pizza = new Pizza();
    }

    // Method with method-chaining to set size
    public PizzaBuilder setSize(Size size) {
        pizza.setSize(size);
        return this;
    }

    // Method with method-chaining to set crust
    public PizzaBuilder setCrust(Crust crust) {
        pizza.setCrust(crust);
        return this;
    }

    // Method with method-chaining to set sauce
    public PizzaBuilder setSauce(Sauce sauce) {
        pizza.setSauce(sauce);
        return this;
    }

    // Method with method-chaining to add a topping
    public PizzaBuilder addTopping(Topping topping) {
        pizza.getToppings().add(topping);
        return this;
    }

    // Method with method-chaining to add many toppings
    public PizzaBuilder addToppings(List<Topping> toppings) {
        pizza.getToppings().addAll(toppings);
        return this;
    }

    // Build method to return the final pizza
    public Pizza build() {
        return pizza;
    }
}

// Test the pizza builder
class Test {
    public static void main(String[] args) {
        PizzaBuilder pizzaBuilder = new PizzaBuilder();

        Pizza myPizza = pizzaBuilder
                .setSize(Size.LARGE)
                .setCrust(Crust.THIN)
                .setSauce(Sauce.TOMATO)
                .addToppings(Arrays.asList(Topping.JALAPENO, Topping.OLIVES))
                .build();

        System.out.println(myPizza);
    }
}